packages/account/src/BiconomySmartAccountV2.ts

```diff
@@ -154,33 +238,277 @@ export class BiconomySmartAccountV2 extends BaseSmartAccount {
+      ...userOp,
+      signature: sig
+    }
+    const signatureWithModuleAddress = ethers.utils.defaultAbiCoder.encode(
+      ['bytes', 'address'],
+      [userOp.signature, await this.activeValidationModule.getAddress()]
+    )
+    userOp.signature = signatureWithModuleAddress
+    return userOp as UserOperation
+  }
+
+  async buildUserOp(
+    transactions: Transaction[],
+    overrides?: Overrides,
+    skipBundlerGasEstimation?: boolean
+  ): Promise<Partial<UserOperation>> {
+    // Review: may not need at all
+    // this.isInitialized()
+
+    // TODO: validate to, value and data fields
+    // TODO: validate overrides if supplied
+    const to = transactions.map((element: Transaction) => element.to)
+    const data = transactions.map((element: Transaction) => element.data ?? '0x')
+    const value = transactions.map((element: Transaction) => element.value ?? BigNumber.from('0'))
+
+    let callData = ''
+    if (transactions.length === 1) {
+      callData = await this.encodeExecute(to[0], value[0], data[0])
+    } else {
+      callData = await this.encodeExecuteBatch(to, value, data)
+    }
+    let nonce = BigNumber.from(0)
+    try {
+      nonce = await this.getNonce()
+    } catch (error) {
+      // Not throwing this error as nonce would be 0 if this.getNonce() throw exception, which is expected flow for undeployed account
+    }
+
+    // let isDeployed = false
+
+    /*if (nonce.eq(0) && this.accountAddress) {
+      isDeployed = await this.isAccountDeployed(this.accountAddress)
+    }*/
+
+    let userOp: Partial<UserOperation> = {
+      sender: this.accountAddress,
+      nonce,
+      initCode: await this.getInitCode(),
+      callData: callData
+    }
+
+    // for this Smart Account current validation module dummy signature will be used to estimate gas
+    userOp.signature = this.getDummySignature()
+
+    userOp = await this.estimateUserOpGas(userOp, overrides, skipBundlerGasEstimation)
+    Logger.log('userOp after estimation ', userOp)
+
+    // Do not populate paymasterAndData as part of buildUserOp as it may not have all necessary details
+    userOp.paymasterAndData = '0x' // await this.getPaymasterAndData(userOp)
+
+    delete userOp.signature
+    return userOp
+  }
+
+  private validateUserOpAndRequest(
+    userOp: Partial<UserOperation>,
+    tokenPaymasterRequest: BiconomyTokenPaymasterRequest
+  ): void {
+    if (!userOp.callData) {
+      throw new Error('Userop callData cannot be undefined')
+    }
+
+    const feeTokenAddress = tokenPaymasterRequest?.feeQuote?.tokenAddress
+    Logger.log('requested fee token is ', feeTokenAddress)
+
+    if (!feeTokenAddress || feeTokenAddress == ethers.constants.AddressZero) {
+      throw new Error(
+        'Invalid or missing token address. Token address must be part of the feeQuote in tokenPaymasterRequest'
+      )
+    }
+
+    const spender = tokenPaymasterRequest?.spender
+    Logger.log('fee token approval to be checked and added for spender: ', spender)
```

Comment:
we are not checking any fee token approval and not adding anything in the request for spender. This log doesn't make sense. Just like above log statement, we can also log here

Logger.log('spender address is ', spender)

Reply:
cool done.